keps/sig-api-machinery/3962-mutating-admission-policies/README.md

```diff
@@ -0,0 +1,1508 @@
+                        ?env: optional.none()
+                    })
+                }
+            }
+        }
+
+```
+
+We will track which fields are unset in this way and remove them after the
+server side apply merge algorithm is run.
+
+This solves the vast majority of value removal needs. Specifically:
+
+| Schema type | Merge type | Example of how to unset a value                                          |
+|-------------| -----------|--------------------------------------------------------------------------|
+| struct      | atomic     | ```Object{ spec: Object.spec{ structField: {?fieldToRemove: optional.none()}}}``` |
+| struct      | granular   | `?fieldToRemove: optional.none()`                                        |
+| map         | atomic     | `mapField: object.spec.mapField.filter(k, k != "keyToRemove")`           |
+| map         | granular   | `mapField: {?"keyToRemove": optional.none()}`                            |
+| list        | atomic     | `listField: object.spec.listField.filter(e, e != "itemToRemove")`        |
+| list        | set        | `setField: object.spec.setField.filter(e, e != "itemToRemove")`          |
+| list        | map        | See below                                                                |
+| list        | granular   | Not Supported                                                            | 
+
+
+List with "map" merge type:
+  - Filter `objects.filter(<list>, <keys-to-remove>)` could be used for the deletion
+  - For associatedList with multiple keys like example above, a directive field added could be used to indicate the deletion.
+```yaml
+mutations:
+  - mutaton: >
+      Object{
+        spec: Object.spec{
+                assocListField: [Object.spec.assocListField{
+                       keyField1: "key1",
+                       keyField2: "key2",
+                       _: optional.none()
+                }]
+        }
+}
+
+```
+
+For examples of removing item from List with Map filtered by a subfield:
+```yaml
+mutations:
+  - mutaton: >
+        Object{
+            spec: Object.spec{
+                containers: object.spec.containers.filter(c, c.envvar != "remove-this-container")
+                }
+            }
+        }
+
+```
+
+For granular list removal, a use case would be removing an item with a sub field named `remove-this-item`.
+```yaml
+mutations:
+  - mutation: >
+      Object{
+        spec: Object.spec{
+                granularList: object.spec.granularList.filter(c, c.subField != "remove-this-item")
+        }
+      }
+```
+
+##### Safety
+
+To ensure mutations are not "broken" by other mutations (overwritten, undone, or
+otherwise invalidated) and ensure the deterministic final state due to the random running order, \
+we provide an option to check if rerun certain mutation policy leads to object change.
+
+Note: this is to ensure the mutation is idempotent in consideration of the random running ordering. 
```

Comment:
Thanks for the review! Yes the idempotent is for a policy itself and it cannot prevent user from writing mutation conflicts with each other. This is an existing problem with mutation webhook so we are not introducing a new issue. The reinvocation mechanism helped in this situation and we plan to follow the existing reinvocation machinist which rerun second time if the object be mutated again. However, it could not prevent additional mutation process to mutate the same field again. For above case of `initContainer` and `container`, I believe there should be certain criteria added and it will stop if there is already a `initContainer` associated with `container` somehow. User should be mindful on the mutation they wanna make to avoid mutual dependency.